core/lib/node/src/node/context.rs

```diff
@@ -0,0 +1,69 @@
+use std::any::Any;
+
+use crate::{node::ZkSyncNode, resource::Resource};
+
+/// An interface to the node's resources provided to the tasks during initialization.
+/// Provides the ability to fetch required resources, and also gives access to the Tokio runtime used by the node.
+#[derive(Debug)]
+pub struct NodeContext<'a> {
```

Comment:
If we made the resource provider into a trait (`trait ResourceProvider<T>`) with a method like `provide(&self) -> T` and parametrized NodeContext and ZkSyncNode by ResourceProvider, then the tasks could accept NodeContext<P> where P : ResourceProvider<whatever the task needs>, and this way resource availability will be resolved at compilation time (although resource caching will still be implemented using Any wrappers.

Reply:
I think I've tried something like this, but met a lot of errors related to object safety (which you get when you replace generics with `dyn Trait`). I'll check once again though.

Reply:
Resource provider traits indeed cannot be generic due to object safety. I've played a bit with attempts to make it more typesafe but failed.
I think I'll keep it as is for now, but we certainly can improve it along the way.